Problem_Set_3/ps3.py

```diff
@@ -0,0 +1,516 @@
+#
+def display_hand(hand):
+    """
+    Displays the letters currently in the hand.
+
+    For example:
+       display_hand({'a':1, 'x':2, 'l':3, 'e':1})
+    Should print out something like:
+       a x x l l l e
+    The order of the letters is unimportant.
+
+    hand: dictionary (string -> int)
+    """
+    
+    for letter in hand.keys():
+        for j in range(hand[letter]):
+             print(letter, end = ' ')      # print all on the same line
+    print()                              # print an empty line
+
+
+# display_hand({'a':1, 'x':2, 'l':3, 'e':1})
+#
+# Make sure you understand how this function works and what it does!
+# You will need to modify this for Problem #4.
+#
+def deal_hand(n):
+    """
+    Returns a random hand containing n lowercase letters.
+    ceil(n/3) letters in the hand should be VOWELS (note,
+    ceil(n/3) means the smallest integer not less than n/3).
+
+    Hands are represented as dictionaries. The keys are
+    letters and the values are the number of times the
+    particular letter is repeated in that hand.
+
+    n: int >= 0
+    returns: dictionary (string -> int)
+    """
+    
+    hand = {'*': 1}
+    num_vowels = int(math.ceil(n / 3)) - 1
+
+    for i in range(num_vowels):
+        x = random.choice(VOWELS)
+        hand[x] = hand.get(x, 0) + 1
+    
+    for i in range(num_vowels, n - 1):    
+        x = random.choice(CONSONANTS)
+        hand[x] = hand.get(x, 0) + 1
+    
+    return hand
+
+# print(deal_hand(7))
+#
+# Problem #2: Update a hand by removing letters
+#
+def update_hand(hand, word):
+    """
+    Does NOT assume that hand contains every letter in word at least as
+    many times as the letter appears in word. Letters in word that don't
+    appear in hand should be ignored. Letters that appear in word more times
+    than in hand should never result in a negative count; instead, set the
+    count in the returned hand to 0 (or remove the letter from the
+    dictionary, depending on how your code is structured). 
+
+    Updates the hand: uses up the letters in the given word
+    and returns the new hand, without those letters in it.
+
+    Has no side effects: does not modify hand.
+
+    word: string
+    hand: dictionary (string -> int)    
+    returns: dictionary (string -> int)
+    """
+    # new_hand = hand
+    # for char in word.lower():
+    #     if char in hand.keys():
+    #         new_hand[char] = hand.get(char, 0) - 1
+    #     else:
+
+    # for char in hand.keys():
+    #     if char in word.lower():
+    #         # if hand[char] > 0:
+    #         new_hand[char] = hand.get(char, 0) - 1
+    #         print(char, new_hand)
+    #     else:
+    #         new_hand[char] = hand.get(char, 0)
+    #         print(char, new_hand)
+    #     if new_hand[char] == 0:
+    #             new_hand.pop(char)
+    # return new_hand
+    new_hand = {}
+    word_dict = get_frequency_dict(word.lower())
+    for char in hand.keys():

```

Comment:
Shouldn't you loop in the opposite way?
i.e loop through all the letters in the word, and see if they exist in the hand?

Reply:
I think I tried doing that initially, but the problem I was running into is that if a letter from the hand wasn't used in the word then I wasn't getting it, cuz I was only looping over the characters in the word. Maybe there's a better way to do it than using the dictionaries like i did, but I don't know what it would be.